source/adapters/level_zero/command_buffer.cpp

```diff
@@ -427,20 +463,51 @@ static ur_result_t enqueueCommandBufferMemCopyRectHelper(
 static ur_result_t enqueueCommandBufferFillHelper(
     ur_command_t CommandType, ur_exp_command_buffer_handle_t CommandBuffer,
     void *Ptr, const void *Pattern, size_t PatternSize, size_t Size,
-    uint32_t NumSyncPointsInWaitList,
+    bool PreferCopyEngine, uint32_t NumSyncPointsInWaitList,
     const ur_exp_command_buffer_sync_point_t *SyncPointWaitList,
     ur_exp_command_buffer_sync_point_t *RetSyncPoint) {
   // Pattern size must be a power of two.
   UR_ASSERT((PatternSize > 0) && ((PatternSize & (PatternSize - 1)) == 0),
             UR_RESULT_ERROR_INVALID_VALUE);
 
-  // Pattern size must fit the compute queue capabilities.
-  UR_ASSERT(
-      PatternSize <=
-          CommandBuffer->Device
-              ->QueueGroup[ur_device_handle_t_::queue_group_info_t::Compute]
-              .ZeProperties.maxMemoryFillPatternSize,
-      UR_RESULT_ERROR_INVALID_VALUE);
+  ze_command_list_handle_t ZeCommandList;
+  // If the copy engine available, the command is enqueued in the
+  // ZeCopyCommandList.
+  if (PreferCopyEngine && CommandBuffer->UseCopyEngine()) {
+    // Check engine fill capabilty (as this is done for non-graph cmd
+    // submission)
+    if (CommandBuffer->Device->hasMainCopyEngine() &&
```

Comment:
When we create the copy command-list, we already check `Device->hasMainCopyEngine() `, so I think we could remove it from this check as we're preconditioned on `UseCopyEngine()`.

Ideally we could also refactor this `if` so that's it's to conditions are copy engine or compute. Rather than having this nested if

Reply:
Good points both, I've removed the extra check and refactored this to be simpler.